### PR TITLE
Bypass task scheduler for reading unsealed pieces

### PR DESCRIPTION
**Pull request by aarshkshah1992**
Picks the minimal set of changes we need from @magik6k's PR at https://github.com/filecoin-project/lotus/pull/6128.

### TODO
- [x] Unit Tests.
  - [x] Test changes to the http handler to check presence of unsealed pieces and to get sector files.
  - [x] Test the `Reader` API on the Remote Store that tries to read an unsealed piece from an already unsealed file on any of the workers.
  - [x] Unit tests for the newly introduced `SectorUnseal` API in the Manager.
  
- [x] Integration Tests.
  - [x] ReadPiece for piece in unsealed file on local worker
  - [x] ReadPiece for piece in sealed file on local worker
  - [x] ReadPiece for piece in unsealed file on remote worker
  - [x] ReadPiece for piece in sealed file on remote worker

- [x] Testing on a devnet/pond with 8MB sector size.
  - [x] Test storage / retrieval
    - Create storage deal
    - Retrieve deal
  - [x] Test storage / retrieval with unsealing
    - Create storage deal
    - Remove unsealed file
    - Retrieve deal (so it unseals)
    - Retrieve deal (so it's freshly unsealed)
  - [x]  Test workers
    - Disable unsealing on miner
    - Enable all tasks for worker (so unsealing must happen on worker)
    - Remove unsealed file
    - Retrieve deal (so it unseals)
    - Retrieve deal (so it's freshly unsealed)

- [x] Manual test against mainnet miner (running single worker)

### CODE REVIEW

File: extern/sector-storage/piece_provider.go
Comment (dirkmc) [inline extern/sector-storage/piece_provider.go:46]:
```suggestion
// Returns a nil reader if the piece does NOT exist in any unsealed file or there is no unsealed file for the given sector on any of the workers.
```

File: extern/sector-storage/piece_provider.go
Comment (dirkmc) [inline extern/sector-storage/piece_provider.go:55]: Suggest adding a comment for exactly what p.storage.Reader does

File: extern/sector-storage/piece_provider.go
Comment (dirkmc) [inline extern/sector-storage/piece_provider.go:60]: Suggest adding a comment for what it means if r == nil

File: extern/sector-storage/piece_provider.go
Comment (dirkmc) [inline extern/sector-storage/piece_provider.go:114]: Do we need to call unlock() before returning the error here?

File: extern/sector-storage/stores/http_handler.go
Comment (dirkmc) [inline extern/sector-storage/stores/http_handler.go:252]:
```suggestion
			log.Error("closing partial file: ", err)
```

File: extern/sector-storage/stores/remote.go
Comment (dirkmc) [inline extern/sector-storage/stores/remote.go:535]:
```suggestion
				// readRemote fetches a reader that we can use to read the unsealed piece from the remote worker.
```

File: extern/sector-storage/stores/remote.go
Comment (dirkmc) [inline extern/sector-storage/stores/remote.go:562]: Love these explanatory comments 👍 👍

File: extern/sector-storage/stores/remote.go
Comment (dirkmc) [inline extern/sector-storage/stores/remote.go:503]: Suggestion for refactor: handle the case where `path != ""` first. It should mean there's less indentation and you don't need the `iloop:` break target anymore

**Review by dirkmc — COMMENTED** [review]
Looking pretty good 👍  - Just a few nits

**Comment by dirkmc** [comment]
I suggest in the description of the PR we add a checkbox list of things that have been done (implement change) and things that need to be done before this PR is ready to be reviewed (adding tests)

File: extern/sector-storage/stores/remote.go
Comment (aarshkshah1992) [inline extern/sector-storage/stores/remote.go:516]: @magik6k 

This looks incorrect, right ? We should sort in descending order of weight because a higher weight means a higher probability that it'll have the unsealed file we are looking for here.

File: extern/sector-storage/piece_provider.go
Comment (aarshkshah1992) [reply extern/sector-storage/piece_provider.go:58]: Done.

File: extern/sector-storage/piece_provider.go
Comment (aarshkshah1992) [reply extern/sector-storage/piece_provider.go:63]: Done.

File: extern/sector-storage/piece_provider.go
Comment (aarshkshah1992) [reply extern/sector-storage/piece_provider.go:122]: Great catch !

**Review by aarshkshah1992 — COMMENTED** [review]
@dirkmc Have addresses all your review comments. Unit Tests are WIP.

File: extern/sector-storage/stores/remote.go
Comment (magik6k) [inline extern/sector-storage/stores/remote.go:549]: Yeah, that's probably a bug, also in acquireFromRemote

Comment (magik6k) [reply extern/sector-storage/stores/remote.go:549]: It shouldn't have much impact, as in most cases sectors are only in one storage, which is likely why it wasn't noticed

**Review by magik6k — COMMENTED** [review]
First review pass, looks good.

We should also bump the minor miner/worker api versions in `api/version.go` given that the storage http api now has the new endpoint

Also ideally we'd test this on a mainnet miner with some workers before merging just to confirm it doesn't have issues in real-world setups

File: extern/sector-storage/storiface/ffi.go
Comment (dirkmc) [inline extern/sector-storage/storiface/ffi.go:23]:
```suggestion
		return xerrors.Errorf("unpadded byte index is %d but must be a multiple of 127", i)
```

**Review by dirkmc — COMMENTED** [review]
Looks good Aarsh 👍 

I'll hold off on approving until we have the tests in place

**Comment by aarshkshah1992** [comment]
@magik6k 

1. We are working on some solids unit tests & integration tests for this PR and you should see them here in some days.
2. Once the unit & integration tests are in place, we'll test it on a devnet with 8K Sectors and on the mainnet too to ensure things are good.

File: extern/sector-storage/piece_provider_test.go
Comment (aarshkshah1992) [inline extern/sector-storage/piece_provider_test.go:53]: @dirkmc 

We should try and use an actual worker here instead of a test worker. You can take a look at how we create an actual worker in one of the worker tests in manager_test.go

Comment (aarshkshah1992) [reply extern/sector-storage/piece_provider_test.go]: @dirkmc I've added commit `3eded3df9dd6edf4e607c7af49026ed995790165` to this PR to use an actual worker here with the actual rust proofs library.

And the tests works fine ! Looking good here :)

File: extern/sector-storage/stores/remote.go
Comment (magik6k) [inline extern/sector-storage/stores/remote.go:595]: Where is this used?

Comment (aarshkshah1992) [reply extern/sector-storage/stores/remote.go:595]: @magik6k 

For unit testing, I needed the `Store` interface to have this method so I could replace a concrete `LocalStore` implementation with a `Store` interface to then mock it out.

Comment (magik6k) [reply extern/sector-storage/stores/remote.go:595]: Can we at least not panic here, just in case this gets called somewhere by accident?

File: extern/sector-storage/piece_provider.go
Comment (nonsense) [inline extern/sector-storage/piece_provider.go:96]: if `r == nil` we don't have to call `unlock` ? Because we do have a few returns in this branch that are not calling `unlock`...?

Comment (dirkmc) [reply extern/sector-storage/piece_provider.go:96]: If r == nil unlock is called within tryReadUnsealedPiece:
https://github.com/filecoin-project/lotus/pull/6280/files/40642b2cad0bb247dd789837c6465163e5b59742#diff-1fe5f948c5ef6d7d11919b998fedd74927b0cfac11ebb3ef1dbaf518fc026999R61

I found this very hard to follow as well, I wonder if there's a way we can refactor to simplify

File: extern/sector-storage/stores/remote.go
Comment (dirkmc) [inline extern/sector-storage/stores/remote.go:568]: I'd suggest we keep track of the last error here.
If we're not able to get the unsealed piece from a remote worker, and there was an error querying a worker, we should return the last error.

File: extern/sector-storage/piece_provider.go
Comment (dirkmc) [inline extern/sector-storage/piece_provider.go:73]: Suggest adding an explanation of the bool return parameter to the godoc

**Review by dirkmc — COMMENTED** [review]
Looking really good Aarsh, I love these comprehensive unit and integration tests 👍 👍

File: extern/sector-storage/stores/remote.go
Comment (aarshkshah1992) [reply extern/sector-storage/stores/remote.go:570]: done.

File: extern/sector-storage/piece_provider.go
Comment (aarshkshah1992) [reply extern/sector-storage/piece_provider.go:73]: done.

File: extern/sector-storage/stores/remote.go
Comment (aarshkshah1992) [reply extern/sector-storage/stores/remote.go]: done.

File: extern/sector-storage/stores/http_handler.go
Comment (aarshkshah1992) [reply extern/sector-storage/stores/http_handler.go]: done.

File: extern/sector-storage/stores/util_unix.go
Comment (aarshkshah1992) [reply extern/sector-storage/stores/util_unix.go:45]: done.

**Comment by aarshkshah1992** [comment]
@magik6k @dirkmc @nonsense Have addressed your review. Please take a look.

**Review by dirkmc — APPROVED** [review]

**Review by magik6k — APPROVED** [review]
Looks great, if testing on mainnet goes well we can probably just merge

**Comment by magik6k** [comment]
`build-lotus-soup` is failing (and not flaky)

**Comment by aarshkshah1992** [comment]
@magik6k The `build-lotus-soup` job has been fixed. Please let's ship this ! :) 🚢

**Review by magik6k — APPROVED** [review]

**Pull request merged**
